pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -115,53 +118,83 @@ func (w *Watcher) registerPlugin(socketPath string) error {
 	if err != nil {
 		return fmt.Errorf("failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
 	}
-	if err := w.invokeRegistrationCallbackAtHandler(ctx, client, infoResp, socketPath); err != nil {
-		return fmt.Errorf("failed to register plugin. Callback handler returned err: %v", err)
-	}
-	glog.V(4).Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", infoResp.Type, infoResp.Name, socketPath)
-	return nil
+
+	return w.invokeRegistrationCallbackAtHandler(ctx, client, infoResp, socketPath)
 }
 
 func (w *Watcher) invokeRegistrationCallbackAtHandler(ctx context.Context, client registerapi.RegistrationClient, infoResp *registerapi.PluginInfo, socketPath string) error {
 	var handlerCbkFn RegisterCallbackFn
 	var ok bool
 	handlerCbkFn, ok = w.handlers[infoResp.Type]
 	if !ok {
+		errStr := fmt.Sprintf("no handler registered for plugin type: %s at socket %s", infoResp.Type, socketPath)
 		if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
 			PluginRegistered: false,
-			Error:            fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", infoResp.Type, socketPath),
+			Error:            errStr,
 		}); err != nil {
-			glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+			return errors.New(errStr + err.Error())
 		}
-		return fmt.Errorf("no handler found registered for plugin type: %s, socket: %s", infoResp.Type, socketPath)
+		return errors.New(errStr)
 	}
 
 	var versions []string
 	for _, version := range infoResp.SupportedVersions {
 		versions = append(versions, version)
 	}
 	// calls handler callback to verify registration request
-	err, chanForAckOfNotification := handlerCbkFn(infoResp.Name, infoResp.Endpoint, versions, socketPath)
+	chanForAckOfNotification, err := handlerCbkFn(infoResp.Name, infoResp.Endpoint, versions, socketPath)
 	if err != nil {
+		errStr := fmt.Sprintf("plugin registration failed with err: %v", err)
 		if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
 			PluginRegistered: false,
-			Error:            fmt.Sprintf("Plugin registration failed with err: %v", err),
+			Error:            errStr,
 		}); err != nil {
-			glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+			return errors.New(errStr + err.Error())
```

Comment:
do you need to `chanForAckOfNotification <- false` before every return? if yes, it might be better to `defer` it. Then a function might be better than a `chan`, easier to defer a function.

Reply:
the bool value written to `chanForAckOfNotification ` is depends on result of `client.NotifyRegistrationStatus`, so i think `defer` will not applicable.